src/node/configuration_tracker.h

```diff
@@ -0,0 +1,413 @@
+          if (nit != nodes.end() && !nit->second.catching_up)
+          {
+            r.emplace(node.first, node.second);
+          }
+        }
+      }
+
+      return r;
+    }
+
+    kv::Configuration::Nodes all_nodes()
+    {
+      kv::Configuration::Nodes r;
+
+      for (const auto& c : configurations)
+      {
+        for (const auto& [id, info] : c.nodes)
+        {
+          r[id] = info;
+        }
+      }
+
+      return r;
+    }
+
+    Index cft_watermark(
+      kv::Version last_idx, std::unordered_map<NodeId, aft::NodeState>& nodes)
+    {
+      LOG_TRACE_FMT(
+        "Configurations: CFT watermark; configurations: {}", to_string());
+      Index r = std::numeric_limits<Index>::max();
+
+      for (auto& c : configurations)
+      {
+        assert(c.nodes.size() > 0);
+
+        // The majority must be checked separately for each active
+        // configuration.
+        std::vector<Index> match;
+        match.reserve(c.nodes.size() + 1);
+
+        for (const auto& node : c.nodes)
+        {
+          auto nit = nodes.find(node.first);
+
+          if (node.first == node_id)
+          {
+            LOG_TRACE_FMT("CFTWM check: {}={}", node_id, last_idx);
+            match.push_back(last_idx);
+          }
+          else if (nit->second.catching_up)
+          {
+            LOG_TRACE_FMT("CFTWM not eligible: {}", nit->first);
+          }
+          else
+          {
+            LOG_TRACE_FMT(
+              "CFTWM check: {}={}", node.first, nit->second.match_idx);
+            match.push_back(nit->second.match_idx);
+          }
+        }
+
+        // `match` may be empty if multiple joining nodes are not caught up and
+        // the current node is leaving, or when the whole network is being
+        // replaced.
+        size_t confirmed = 0;
+        if (!match.empty())
+        {
+          sort(match.begin(), match.end());
+          confirmed = match.at((match.size() - 1) / 2);
+          r = std::min(confirmed, r);
+        }
+
+#ifndef NDEBUG
+        std::stringstream ss;
+        for (auto& m : match)
+          ss << m << " ";
+        LOG_TRACE_FMT(
+          "last_idx={} confirmed={} r={} match={}",
+          last_idx,
+          confirmed,
+          r,
+          ss.str());
+#endif
+      }
+
+      return r;
+    }
+
+    std::string to_string() const
+    {
+      std::stringstream ss;
+      for (auto c : configurations)
+      {
+        bool first = true;
+        ss << "[";
+        for (const auto& [id, _] : c.nodes)
+        {
+          if (first)
+            first = false;
+          else
+            ss << " ";
+          ss << id;
+          auto nit = nodes.find(id);
+          if (nit == nodes.end() && id != node_id)
+            ss << "!";
+          else if (nit != nodes.end() && nit->second.catching_up)
+            ss << "*";
+        }
+        ss << "]@" << c.idx << " ";
+      }
+      return ss.str();
+    }
+
+    bool record_promotion(const NodeId& node_id)
+    {
```

Comment:
Sorry i might be misunderstanding what is going on here, why does this need to go through the frontend?

Reply:
Recommendation by @achamayou. This makes sure the transaction is properly recorded on the ledger.